src/mpid/common/shm/mpidu_shm_alloc.c

```diff
@@ -0,0 +1,702 @@
+    MPIR_ERR_CHECK(mpi_errno);
+
+    *maxsz_loc = maxloc_result.loc;
+    *maxsz = (size_t) maxloc_result.sz;
+
+  fn_exit:
+    if (maxloc_type != MPI_DATATYPE_NULL)
+        MPIR_Type_free_impl(&maxloc_type);
+    if (maxloc_op != MPI_OP_NULL)
+        MPIR_Op_free_impl(&maxloc_op);
+    return mpi_errno;
+  fn_fail:
+    goto fn_exit;
+}
+
+/* Try mapping shmem into processes address space symmetrically using address passed in
+ * through shm_seg->base_addr. Processes in different nodes coordinate after the function
+ * returns. */
+static int map_symm_shm(MPIR_Comm * shm_comm_ptr, MPIDU_shm_seg_t * shm_seg, int *map_result_ptr)
+{
+    int mpi_errno = MPI_SUCCESS, mpl_err = 0;
+    int all_map_result = SYMSHM_MAP_FAIL;
+    bool mapped_flag = false;
+    MPIR_Errflag_t errflag = MPIR_ERR_NONE;
+
+    if (shm_seg->segment_len > 0) {
+        if (shm_comm_ptr) {
+            if (shm_comm_ptr->rank == 0) {
+                char *serialized_hnd = NULL;
+
+                /* root prepare shm segment */
+                mpl_err = MPL_shm_fixed_seg_create_and_attach(shm_seg->hnd, shm_seg->segment_len,
+                                                              (void **) &(shm_seg->base_addr), 0);
+                if (mpl_err) {
+                    *map_result_ptr =
+                        (mpl_err == MPL_SHM_EINVAL) ? SYMSHM_MAP_FAIL : SYMSHM_OTHER_FAIL;
+                    goto root_sync;
+                } else
+                    mapped_flag = true;
+
+                mpl_err = MPL_shm_hnd_get_serialized_by_ref(shm_seg->hnd, &serialized_hnd);
+                if (mpl_err) {
+                    *map_result_ptr = SYMSHM_OTHER_FAIL;
+                }
+
+              root_sync:
+                /* broadcast the mapping result on rank 0 */
+                mpi_errno = MPIR_Bcast(map_result_ptr, 1, MPI_INT, 0, shm_comm_ptr, &errflag);
+                MPIR_ERR_CHKANDJUMP(errflag, mpi_errno, MPI_ERR_OTHER, "**coll_fail");
+
+                if (*map_result_ptr != SYMSHM_SUCCESS)
+                    goto map_fail;
+
+                mpi_errno =
+                    MPIR_Bcast(serialized_hnd, MPL_SHM_GHND_SZ, MPI_BYTE, 0,
+                               shm_comm_ptr, &errflag);
+                MPIR_ERR_CHKANDJUMP(errflag, mpi_errno, MPI_ERR_OTHER, "**coll_fail");
+
+                /* ensure all other processes have mapped successfully */
+                mpi_errno = MPIR_Allreduce(map_result_ptr, &all_map_result, 1, MPI_INT,
+                                           MPI_MAX, shm_comm_ptr, &errflag);
+                MPIR_ERR_CHKANDJUMP(errflag, mpi_errno, MPI_ERR_OTHER, "**coll_fail");
+
+                if (all_map_result != SYMSHM_SUCCESS)
+                    goto map_fail;
+            } else {
+                char serialized_hnd[MPL_SHM_GHND_SZ] = { 0 };
+
+                mpi_errno = MPIR_Bcast(map_result_ptr, 1, MPI_INT, 0, shm_comm_ptr, &errflag);
```

Comment:
Can you please add back the comments for non-root process's code? Please see the original code. The code is complex, let us keep all necessary comments.

Reply:
done